### ISSUE TITLE
Add support for more data types

### ISSUE THREAD

**Issue by cjpatton** [abetterinternet/libprio-rs#23]
This library currently supports a single input type, a vector of 0s and 1s. In [Prio v3](https://github.com/abetterinternet/prio-documents) we will need support for a much richer set of input types. In my opinion, we ought to aim for the full range of input types that [[BBG+19](https://eprint.iacr.org/2021/017), Theorem 4.3] supports. In order to maximize code reuse, what I suggest we do is implement the core logic of the fully linear PCP system and delegate the implementation of the validity circuit to each data type. Something like this (e.g., `pcp.rs`):
```rust

//! This module implements the fully linear PCP ("Probabilistically Checkable Proof") system
//! decsribed in [[BBG+19](https://eprint.iacr.org/2021/017), Theorem 4.3]. It is the core
//! component of Prio's input-validation protocol.
//!
//! A PCP asserts that a given input is a member of a formal language, e.g., that each integer in
//! an input vector is within a certain range. The proof system is comprised of a proover and a
//! verifier. The prover generates a PCP for its input, and the verifier checks the PCP to
//! determine if the input is valid. This procedure should always succeed when the input is valid;
//! if the input is invalid, then the validatin procedure should fail, except with small
//! probability.
//!
//! A "fully linear" PCP system is a PCP system that can operate on secret shared data. In this
//! setting, the prover splits its input and proof into shares and distributes the shares among a
//! set of verifiers. Th verifiers then run a protocol in which they learn whether the input is
//! valid, but without learning the input itself.
//!
//! The fully linear PCP system of [BBG+19, Theorem 4.3] applies languages recognized by arithmetic
//! circuits over finite fields that have a particular structure. Namely, all gates in the circuit
//! are either affine (i.e., addition or scalar multiplication) or invoke a special sub-circuit,
//! called the gadget, which may contain non-affice operations (i.e., multiplication).

use crate::field::FieldElement;

/// This type represents an input to be validated.
pub trait Datum<F: FieldElement, G: Gadget<F>>: Sized {
    /// Evalauates the arithmetic circuit on the given input (i.e., `self`) and returns the output.
    /// `rand` is the random input of the validity circuit.
    fn valid(&self, g: &mut dyn Gadget<F>, rand: &[F]) -> F;

    /// Returns an instance of gadget associated with the validity circuit. The length of the proof
    /// generated for this data type is linear in the number of times the gadget is invoked.
    fn gadget(&self) -> G;

    /// Returns a reference to the input encoded as a vector of field elements.
    fn vec(&self) -> &[F];

    /// Tries to construct an instance of this type from a vector of field Elements. Returns `None`
    /// if the input is not the correct length.
    fn from_vec(vec: &[F]) -> Option<Self>;
}

/// The sub-circuit associated with some validity circuit.
pub trait Gadget<F: FieldElement> {
    /// Evaluates the gadget on input `inp` and returns the output.
    fn call(&self, inp: &[F]) -> F;
}

/// Generate a PCP of the validity of `x`'. This is algorithm is run by the prover.
pub fn prove<F: FieldElement, G: Gadget<F>, T: Datum<F, G>>(x: &T) -> Proof<F> {
    // TODO
}

/// The output of `prove`.
pub struct Proof<F: FieldElement> {
    // TODO
}

/// Generate the verification message for input `x` and proof `pf`, and randomness `rand`. This
/// algorithm is run by the verifier. In Prio, each aggregator runs this algorithm on a share of
/// the proof and input.
pub fn query<F: FieldElement, G: Gadget<F>, T: Datum<F, G>>(
    x: &T,
    pf: &Proof<F>,
    rand: &[F],
) -> Verifier<F> {
    // TODO
}

/// The output of `query`.
pub struct Verifier<F: FieldElement> {
    // TODO
}

/// Decide if input `x` is valid based on verification message `vf`. This algorithm is run by the
/// verifier. In Prio, this algorithm is run by the leader and the output is distributed among the
/// rest of the aggregators.
pub fn decide<F: FieldElement, G: Gadget<F>, T: Datum<F, G>>(x: &T, vf: &Verifier<F>) -> bool {
    // TODO
}
```

`prove`, `query`, and `decide` correspond to the three algorithms described in {{protocol}} of the [design doc](https://github.com/abetterinternet/prio-documents/blob/main/design-document.md). For example:
```rust
let pf = prove(x); // Run by the prover
...
let share_of_vf = query(share_of_x, share_of_pf); // Run by each of the aggregators
...
let b = decide(share_of_x, vf); // Run by the leader
```

**Comment by cjpatton** [comment]
Based on today's design meeting, I will re-work this API to make a distinction between input (resp. proof) and input share (resp. proof share).

**Comment by degregat** [comment]
Thats great news! I'd be happy to contribute fixed point types once its done (see #9 ).

**Comment by cjpatton** [comment]
Now that we've landed the PCP framework code, we should be able to support whatever proof system follows from [BBG+19, Theorem 4.3].
